stats/stats.go

```diff
@@ -0,0 +1,146 @@
+package stats
+
+import (
+	"context"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/elliotchance/orderedmap"
+	"github.com/hashicorp/go-hclog"
+	"github.com/segmentio/stats/v4"
+)
+
+type stat struct {
+	start    time.Time
+	duration time.Duration
+	stopped  bool
+}
+
+type durationLogger struct {
+	logger            hclog.Logger
+	trackedOperations *orderedmap.OrderedMap
+	mu                sync.Mutex
+}
+
+type Options struct {
+	tick    time.Duration
+	handler stats.Handler
+}
+
+func NewClockWithObserve(name string, tags ...stats.Tag) *stats.Clock {
+	// The default clock doesn't send a measurement on start (only on stop)
+	// We want both on start AND stop, so we wrap the ClockAt method
+	cl := stats.DefaultEngine.ClockAt(name, time.Now(), tags...)
+	stats.DefaultEngine.Observe(name, time.Duration(0), tags...)
+	return cl
+}
+
+// This is executed in the context of the calling method
+// We would like to keep track of still running operations, and completed operations durations
+// HandleMeasures can be called by `NewClockWithObserve` which indicates a "start" of an operation
+// Or by `clock.Stop` which indicates a "stop" of an operation
+// We pass the measurements to a channel and periodically aggregate the data and print a hearbeat log
+func (h *durationLogger) HandleMeasures(time time.Time, measures ...stats.Measure) {
+	h.mu.Lock()
+	defer h.mu.Unlock()
+	for _, m := range measures {
+		id, stamp := getMeasurementDetails(m.Fields[0].Name, m.Tags)
+		if stamp {
+			item, ok := h.trackedOperations.Get(id)
+			if ok {
+				h.trackedOperations.Set(id, stat{start: item.(stat).start, duration: m.Fields[0].Value.Duration(), stopped: true})
+			}
+		} else {
+			h.trackedOperations.Set(id, stat{start: time, duration: 0, stopped: false})
```

Comment:
Not sure what is the style we want. We could use https://github.com/GaijinEntertainment/go-exhaustruct to enforce initializing structs

Reply:
no need, zero value is valid in Go.
```suggestion
			h.trackedOperations.Set(id, stat{start: time})
```